pkg/backends/calico/routes.go

```diff
@@ -195,24 +197,104 @@ func (rg *routeGenerator) setRouteForSvc(svc *v1.Service, ep *v1.Endpoints) {
 	// see if any endpoints are on this node and advertise if so
 	// else remove the route if it also already exists
 	rg.Lock()
-	if rg.advertiseThisService(svc, ep) {
+	defer rg.Unlock()
+
+	advertise := rg.advertiseThisService(svc, ep)
+	if advertise {
 		route := svc.Spec.ClusterIP + "/32"
-		if cur, exists := rg.svcRouteMap[key]; !exists {
+		if cur, exists := rg.svcClusterRouteMap[key]; !exists {
 			// This is a new route - send it through.
-			rg.advertiseRoute(key, route)
+			rg.advertiseClusterRoute(key, route)
 		} else if route != cur {
 			// The route has changed somehow. Send a delete for the old one
 			// and add the new one. We don't expect to hit this, since cluster IPs
 			// are immutable. But, handle it for good measure.
 			logc.Warn("ClusterIP for service changed, adjusting")
-			rg.withdrawRoute(key, cur)
-			rg.advertiseRoute(key, route)
+			rg.withdrawClusterRoute(key, cur)
+			rg.advertiseClusterRoute(key, route)
 		}
-	} else if cur, exists := rg.svcRouteMap[key]; exists {
+	} else if cur, exists := rg.svcClusterRouteMap[key]; exists {
 		// We were advertising this route, but should no longer do so.
-		rg.withdrawRoute(key, cur)
+		rg.withdrawClusterRoute(key, cur)
+	}
+
+	// External IP's
+	if advertise {
+
+		advertisedExternalRoutes := rg.svcExternalRouteMap[key]
+		if advertisedExternalRoutes == nil {
+			advertisedExternalRoutes = make(map[string]bool)
+		}
+
+		externalRoutes := addSuffix(svc.Spec.ExternalIPs, "/32")
+
+		// Advertise any External IP's which we don't already advertise
+		for _, route := range externalRoutes {
+
+			// Only advertise whitelisted External IP's
+			if !rg.isAllowedExternalRoute(route) {
```

Comment:
I think this code handles the case where we receive the BGPConfiguration first, followed by a series of Services.

However, I'm not sure it handles the case where we receive the services first, followed by an update to the BGP configuration.

I think we'll need some way to kick this code when the set of allowed external IP cidrs we get from the syncer changes. For example, if I remove one of the allowed CIDRs from the BGPConfiguration, we need this code to withdraw any routes it has previously advertised.

Reply:
One way we could accomplish this would be a function which checks if anything in the cache is now invalid. `updateExternalIPNets` could then trigger this function whenever the syncer updates the list.

Reply:
Yeah, we probably want another function `onBGPConfigurationUpdate` that does the following:
- Iterates through all services we know about
- For each, advertise any external IPs that are within the CIDRs (if they already are advertised, this is a no-op)
- For each, withdraw any external IPs that are not within the CIDRs (if they already are withdrawn, this is a no-op)

Then, we can call this as part of the sync processing when we get notified of a BGPConfig change.

Reply:
This is now done - I have a `onBGPConfigurationUpdate` function that is called when the global serviceExternalIPs are changed.